strawberryfields/api_client.py

```diff
@@ -0,0 +1,546 @@
+        and password authentication and calls set_authorization_header.
+
+        Args:
+            username (str): A user name.
+            password (str): A password.
+        """
+        raise NotImplementedError()
+
+    def set_authorization_header(self, authentication_token):
+        """
+        Adds the authorization header to the headers dictionary to be included
+        with all API requests.
+
+        Args:
+            authentication_token (str): An authentication token used to access the API.
+        """
+        self.headers["Authorization"] = authentication_token
+
+    def join_path(self, path):
+        """
+        Joins a base url with an additional path (e.g. a resource name and ID)
+
+        Args:
+            path (str): A path to be joined with BASE_URL.
+
+        Returns:
+            str: A joined path.
+        """
+        return join_path(self.BASE_URL, path)
+
+    def get(self, path):
+        """
+        Sends a GET request to the provided path. Returns a response object.
+
+        Args:
+            path (str): A path to send the GET request to.
+
+        Returns:
+            requests.Response: A response object, or None if no response could be fetched from the
+            server.
+        """
+        try:
+            response = requests.get(url=self.join_path(path), headers=self.HEADERS)
+        except requests.exceptions.ConnectionError as e:
+            response = None
+            warnings.warn("Could not connect to server ({})".format(e))
+        return response
+
+    def post(self, path, payload):
+        """
+        Converts payload to a JSON string. Sends a POST request to the provided
+        path. Returns a response object.
+
+        Args:
+            path (str): A path to send the GET request to.
+            payload: A JSON serializable object to be sent to the server.
+
+        Returns:
+            requests.Response: A response object, or None if no response could be fetched from the
+            server.
+        """
+        # TODO: catch any exceptions from dumping JSON
+        data = json.dumps(payload)
+        try:
+            response = requests.post(url=self.join_path(path), headers=self.HEADERS, data=data)
+        except requests.exceptions.ConnectionError as e:
+            response = None
+            warnings.warn("Could not connect to server ({})".format(e))
+        return response
+
+
+class ResourceManager:
+    """
+    This class handles all interactions with APIClient by the resource.
+    """
+
+    http_status_code = None
+
+    def __init__(self, resource, client=None):
+        """
+        Initialize the manager with resource and client instances. A client
+        instance is used as a persistent HTTP communications object, and a
+        resource instance corresponds to a particular type of resource (e.g.
+        Job)
+        """
+        setattr(self, "resource", resource)
+        setattr(self, "client", client or APIClient())
```

Comment:
How come you're using `setattr` here, as opposed to the following?
```python
self.resource = resource
```

Reply:
I think this is a remnant of having had keys in a tuple earlier. There is no reason to keep them like this at this point and I prefer just straight assignment. I will include this change in a clean up commit.